ml4gw/waveforms/taylorf2.py

```diff
@@ -52,13 +58,60 @@ def taylorf2_phase(
     pfa_v1 = 0.0
     pfa_v2 = 5.0 * (74.3 / 8.4 + 11.0 * eta) / 9.0
     pfa_v3 = -16.0 * PI
+    # SO contributions at 1.5 PN
+    pfa_v3 += (
+        m1byM * (25.0 + 38.0 / 3.0 * m1byM) * chi1
+        + m2byM * (25.0 + 38.0 / 3.0 * m2byM) * chi2
+    )
     pfa_v4 = (
         5.0
         * (3058.673 / 7.056 + 5429.0 / 7.0 * eta + 617.0 * eta * eta)
         / 72.0
     )
+    # SO, SS, S1,2-squared contributions
+    pfa_v4 += (
+        247.0 / 4.8 * eta * chi1 * chi2
+        + -721.0 / 4.8 * eta * chi1 * chi2
+        + (-720.0 / 9.6 * m1byM * m1byM + 1.0 / 9.6 * m1byM * m1byM) * chi1sq
+        + (-720.0 / 9.6 * m2byM * m2byM + 1.0 / 9.6 * m2byM * m2byM) * chi2sq
+        + (240.0 / 9.6 * m1byM * m1byM + -7.0 / 9.6 * m1byM * m1byM) * chi1sq
+        + (240.0 / 9.6 * m2byM * m2byM + -7.0 / 9.6 * m2byM * m2byM) * chi2sq
+    )
     pfa_v5logv = 5.0 / 3.0 * (772.9 / 8.4 - 13.0 * eta) * PI
     pfa_v5 = 5.0 / 9.0 * (772.9 / 8.4 - 13.0 * eta) * PI
+    # SO coefficient for 2.5 PN
+    pfa_v5logv += 3.0 * (
+        -m1byM
+        * (
+            1391.5 / 8.4
+            - m1byM * (1.0 - m1byM) * 10.0 / 3.0
```

Comment:
Sorry this is too exciting of a development and I had to take a quick peek. My only two cents I'll throw in is that, depending on where your bottlenecks are, it could be worth being deliberate about how you include constants here. For example, this will first compute `m1byM * (1.0 - m1byM)`, which will be a tensor, then multiply that element-wise by 10, then divide that tensor element-wise by 3. So you're paying for an extra kernel launch and element-wise multiply that you wouldn't need if you did `10.0 / 3.0. * m1byM * (1.0 - m1byM)`. A small thing, but could add up in the aggregate, especially for some of the square terms you could pre-compute. Anyway, minor point, this is extremely cool.

Reply:
Oh interesting, yeah, we should do a timing comparison and see what sort of speed up we can get. Could be worth deviating from lalsuite for.

Reply:
OK I did some quick timing test locally. With lalsimulation, getting waveforms in sequence takes around 3 milliseconds per waveform generation. With the torch implementation, a batch of 1000 waveforms is created in 500 ms i.e. 0.5 ms per waveform. This is on my laptop that does not have a GPU.

```
%%timeit -n 100
hp1, hc1 = lalsimulation.SimInspiralChooseFDWaveform(**pars)

2.9 ms ± 275 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

vs.

```
params = torch.tensor([m1, m2, chi1, chi2, D, phic, inclination])

params = params.repeat([1000, 1])
params.shape # torch.Size([1000, 7])

%%time
hp_torch, hc_torch = taylorf2.TaylorF2(
    torch_freqs,
    params[:,0],
    params[:,1],
    params[:,2],
    params[:,3],
    params[:,4],
    params[:,5],
    params[:,6],
    f_ref
)

CPU times: user 4.55 s, sys: 2.12 s, total: 6.67 s
Wall time: 511 ms
```

Reply:
I moved the scalar pre-factors ahead.

Reply:
Nice! I tried computing all the repeated tensor products up front, and saw a small improvement there as well. However, it involved a lot more rearranging, and I think it's not worth deviating much from lalsim at this point.

Reply:
Interesting, yeah 10% seems about appropriate, though you might observe a bit more acceleration on GPU where the compute time is fast enough to be comparable to the kernel launch overhead. But either way, insanely stoked to see what you guys do with this.